platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/registry/PrometheusPushMeterRegistry.java

```diff
@@ -0,0 +1,86 @@
+package org.hypertrace.core.serviceframework.metrics.registry;
+
+import io.micrometer.core.instrument.util.TimeUtils;
+import io.micrometer.core.lang.Nullable;
+import io.micrometer.prometheus.PrometheusMeterRegistry;
+import io.prometheus.client.exporter.PushGateway;
+import java.io.IOException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import org.hypertrace.core.serviceframework.metrics.PlatformMetricsRegistry;
+import org.hypertrace.core.serviceframework.metrics.config.PrometheusPushRegistryConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Metric Registry for Prometheus Push Gateway
+ */
+public class PrometheusPushMeterRegistry extends PrometheusMeterRegistry {
+  private static final Logger logger = LoggerFactory.getLogger(PlatformMetricsRegistry.class);
+  private final PrometheusPushRegistryConfig pushConfig;
+  private final PushGateway pushGateway;
+
+  @Nullable
+  private ScheduledExecutorService scheduledExecutorService;
+
+  public PrometheusPushMeterRegistry(
+      PrometheusPushRegistryConfig pushConfig, ThreadFactory threadFactory, PushGateway pushGateway) {
+    super(pushConfig::get);
+
+    pushConfig.requireValid();
+
+    this.pushConfig = pushConfig;
+    this.pushGateway = pushGateway;
+
+    start(threadFactory);
+  }
+
+  public void publish() throws IOException {
+      pushGateway.pushAdd(getPrometheusRegistry(), pushConfig.jobName());
+  }
+
+  /**
+   * Catch uncaught exceptions thrown from {@link #publish()}.
+   */
+  private void publishSafely() {
+    try {
+      publish();
+    } catch (Throwable e) {
+      logger.warn("Unexpected exception thrown while publishing metrics for " + this.getClass()
+          .getSimpleName(), e);
+    }
+  }
+
+  public void start(ThreadFactory threadFactory) {
+    if (scheduledExecutorService != null)
+      stop();
+
+    if (pushConfig.enabled()) {
+      logger.info("publishing metrics for " + this.getClass().getSimpleName() + " every " + TimeUtils
+          .format(pushConfig.step()));
+
+      scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(threadFactory);
+      scheduledExecutorService.scheduleAtFixedRate(this::publishSafely, pushConfig.step()
```

Comment:
I'm a little confused. So this periodic push is from the push-gateway to prometheus?

If that is the case, how is it different from the periodic poll from prometheus? The last set of metrics before the application dies is what we need here right?
May be i'm missing something

Reply:
Yes, the periodic push is from service to push-gateway. Then, Prometheus can scrap from push-gateway.

The rate at which a job pushes metric can be decoupled than the rate when Prometheus scrape. If seconds is not granular enough, we can make it milliseconds and have finer control. 

Do we have jobs that runs faster than milliseconds that requires push on-demand? If we do, then we can expose getting the registry from service-platform and do on-demand publish.

Reply:
The current api-naming trainer which is configured to run every 5 min, finishes in less than a minute.
So my point is not about configuring the granularity(because no matter what, there will be something missing). But rather, when the application stops(for which we already have a hook in our framework), there should be a push performed irrespective of whatever the granularity is.

Reply:
Similar to how we currently call `consoleReporter.report();` in the stop method

```
private static void stopConsoleMetricsReporter() {
    if (consoleReporter == null) {
      return;
    }
    // report all metrics one final time
    consoleReporter.report();
    // stop console reporter
    consoleReporter.stop();
  }
```

Reply:
Makes sense. Push data before the entire job closes down. 

I wonder if we should've called the stop() for each of the METRIC_REGISTRY.